odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+
+        self.alpha = alpha
+        self.TV = TV(domain, grad=grad)
+        self.grad = self.TV.right
+        self.NN = odl.solvers.IndicatorBox(domain, 0, np.inf)
+        self.L2 = 0.5 * odl.solvers.L2NormSquared(domain)
+
+        super().__init__(space=domain, linear=False, grad_lipschitz=0)
+
+    def __call__(self, x):
+        """ Characteristic function of the non-negative orthant
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the total variation of a constant is zero
+
+        >>> import odl.contrib.spdhg as spdhg, odl
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = space.one()
+        >>> TVNN(x) < 1e-10
+
+        Check that negative functions are mapped to infty
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> np.isinf(TVNN(x))
+        """
+
+        nn = self.NN(x)
+
+        if nn is np.inf:
+            return nn
+        else:
+            out = self.alpha * self.TV(x) + nn
+            if self.strong_convexity > 0:
+                out += self.strong_convexity * self.L2(x)
+            return out
+
+    def proximal(self, sigma):
+        """ Prox operator of TV. It allows the proximal step length to be a vector
+        of positive elements.
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the proximal operator is the identity for sigma=0
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> y = TVNN.proximal(0)(x)
+        >>> (y-x).norm() < 1e-10
+
+        Check that negative functions are mapped to 0
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> y = TVNN.proximal(0.1)(x)
+        >>> y.norm() < 1e-10
+        """
+
+        if sigma == 0:
+            return odl.IdentityOperator(self.domain)
+
+        else:
+            def TV_prox(z, out=None):
+
+                if out is None:
+                    out = z.space.zero()
+
+                opts = self.prox_options
+
+                sigma_ = np.copy(sigma)
+                z_ = z.copy()
```

Comment:
Do this in the `else` case a bit down. I also don't see a reason to copy `sigma`. The old values are never used.

Reply:
I don't want to modify the old values which happens in the next lines. The code is intended to be used for both scalar and vector-valued sigma.

Reply:
Okay, good point. You can also alias the old name by doing `sigma = <some computation involving sigma>`, that also doesn't change the input and uses one less new name.